developer-hub/src/reference_docs/defensive_programming.rs

```diff
@@ -0,0 +1,558 @@
+//! As our runtime should _never_ panic; this includes carefully handling [`Result`]/[`Option`]
+//! types, eliminating the possibility of integer overflows, converting between number types, or
+//! even handling floating point usage with fixed point arithmetic to mitigate issues that come with
+//! floating point calculations.
+//!
+//! Intentional and predictable design should be our first and foremost
+//! property for ensuring a well running, safely designed system.
+//!
+//! ## Defensive Programming
+//!
+//! Defensive programming is a design paradigm that enables a particular program to continue
+//! running despite unexpected behavior. These unforseen circumstances may
+//! cause the program to stop, (or in the Rust context, `panic!`) defensive practices allow for
+//! these circumstances to be accounted for ahead of time, and for them to be handled in a more
+//! graceful manner.
+//!
+//! The Polkadot SDK is both built to reflect these principles, and to be used accordingly.
+//!
+//! ## General Practices
+//!
+//! When developing within the context of the a runtime, there is one golden rule:
+//!
+//! ***DO NOT PANIC,*** most of the time. There are some exceptions, such as critical operations being actually more
+//! dangerous than allowing the node to continue running (block authoring, consensus, etc).
+//!
+//!  General guidelines:
+//!
+//! - Avoid panickable operations, such as directly using `unwrap()` for a [`Result`], common errors such as accessing collections out of bounds (using safer methods to access collection types, i.e., `get()`).
+//! - It may be acceptable to use `except()`, but only if one is completely certain (and has
+//!   performed a check beforehand) that a value won't panic upon unwrapping.
+//! - If you are writing a function that could panic, [be sure to document it!](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html#documenting-components)
+//! - Carefully handle mathematical operations.  Many seemingly, simplistic operations, such as **arithmetic** in the runtime, could present a
+//!   number of issues [(see more later in this document)](#integer-overflow).
+//!
+//! ### Defensive Traits
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait provides a number of functions, all of which
+//! provide an alternative to 'vanilla' Rust functions, e.g.,:
+//!
+//! - [`defensive_unwrap_or()`](frame::traits::Defensive::defensive_unwrap_or)
+//! - [`defensive_ok_or()`](frame::traits::DefensiveOption::defensive_ok_or)
+//!
+//! The [`Defensive`](frame::traits::Defensive) trait and its companions,
+//! [`DefensiveOption`](frame::traits::DefensiveOption),
+//! [`DefensiveResult`](frame::traits::DefensiveResult) can be used to defensively unwrap
```

Comment:
I dont know if we have them listed somewhere, but there are a few more: ![Screenshot 2024-01-02 at 11 38 07](https://github.com/paritytech/polkadot-sdk/assets/10380170/e3c0a7e1-0133-4963-b756-2af4a73a4f5f)